src/ripple/app/ledger/impl/LedgerReplayTask.cpp

```diff
@@ -0,0 +1,271 @@
+namespace ripple {
+
+LedgerReplayTask::TaskParameter::TaskParameter(
+    InboundLedger::Reason r,
+    uint256 const& finishLedgerHash,
+    std::uint32_t totalNumLedgers)
+    : reason(r), finishHash(finishLedgerHash), totalLedgers(totalNumLedgers)
+{
+    assert(finishLedgerHash.isNonZero() && totalNumLedgers > 0);
+}
+
+bool
+LedgerReplayTask::TaskParameter::update(
+    uint256 const& hash,
+    std::uint32_t seq,
+    std::vector<uint256> const& sList)
+{
+    if (finishHash != hash || sList.size() + 1 < totalLedgers || full)
+        return false;
+
+    finishSeq = seq;
+    skipList = sList;
+    skipList.emplace_back(finishHash);
+    startHash = skipList[skipList.size() - totalLedgers];
+    assert(startHash.isNonZero());
+    startSeq = finishSeq - totalLedgers + 1;
+    full = true;
+    return true;
+}
+
+bool
+LedgerReplayTask::TaskParameter::canMergeInto(
+    TaskParameter const& existingTask) const
+{
+    if (reason == existingTask.reason)
+    {
+        if (finishHash == existingTask.finishHash &&
+            totalLedgers <= existingTask.totalLedgers)
+        {
+            return true;
+        }
+
+        if (existingTask.full)
+        {
+            auto const& exList = existingTask.skipList;
+            if (auto i = std::find(exList.begin(), exList.end(), finishHash);
+                i != exList.end())
+            {
+                return existingTask.totalLedgers >=
+                    totalLedgers + (exList.end() - i) - 1;
+            }
+        }
+    }
+
+    return false;
+}
+
+LedgerReplayTask::LedgerReplayTask(
+    Application& app,
+    InboundLedgers& inboundLedgers,
+    LedgerReplayer& replayer,
+    std::shared_ptr<SkipListAcquire>& skipListAcquirer,
+    TaskParameter&& parameter)
+    : TimeoutCounter(
+          app,
+          parameter.finishHash,
+          LedgerReplayParameters::TASK_TIMEOUT,
+          {jtREPLAY_TASK,
+           "LedgerReplayTask",
+           LedgerReplayParameters::MAX_QUEUED_TASKS},
+          app.journal("LedgerReplayTask"))
+    , inboundLedgers_(inboundLedgers)
+    , replayer_(replayer)
+    , parameter_(parameter)
+    , maxTimeouts_(std::max(
+          LedgerReplayParameters::TASK_MAX_TIMEOUTS_MINIMUM,
+          parameter.totalLedgers *
+              LedgerReplayParameters::TASK_MAX_TIMEOUTS_MULTIPLIER))
+    , skipListAcquirer_(skipListAcquirer)
+{
+    JLOG(m_journal.trace()) << "Task ctor " << mHash;
+}
+
+LedgerReplayTask::~LedgerReplayTask()
+{
+    JLOG(m_journal.trace()) << "Task dtor " << mHash;
+}
+
+void
+LedgerReplayTask::init()
+{
+    JLOG(m_journal.debug()) << "Task start " << mHash;
+    ScopedLockType sl(mLock);
+    trigger(sl);
+    if (!isDone())
+        setTimer();
+}
+
+void
+LedgerReplayTask::trigger(ScopedLockType& sl)
+{
+    JLOG(m_journal.trace()) << "trigger " << mHash;
+    if (isDone() || !parameter_.full)
+        return;
+
+    if (!parent_)
+    {
+        parent_ = app_.getLedgerMaster().getLedgerByHash(parameter_.startHash);
```

Comment:
I think `startHash` can be zero here. Consider this sequence of events in pseudocode:

```cpp
LedgerReplayer::replay() {
  skipList = new SkipListAcquire(...);
  task = new LedgerReplayTask(...);
  assert skipList.addTask(task) == true;
  task.init()
}
```

If we initialize the task before the skip list has returned, how could we possibly know the start ledger ID? It must be empty.

When that happens, I think that `LedgerMaster::getLedgerByHash` returns `nullptr`, but could not verify on a quick check. However, `InboundLedgers::acquire` _asserts_ that the ledger ID is not zero.

`InboundLedgers` does not let us register a callback to notify this task when the start ledger has arrived. Instead, this task checks upon each timeout. I think we need an early exit if `parameter_.startHash` is zero.

Reply:
Oh, this is guarded by the check on line 130 for `parameter_.full`. Never mind!